akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala

```diff
@@ -444,108 +568,173 @@ private[akka] class Shard(
+              // FIXME should trigger backoff + restart?
+              log.warning("Stop of [{}] arrived, should lead to backoff restart, not implemented yet.", entityId)
+            case Passivating(_) =>
+              // will go in next batch update
+              log.debug("Stop of [{}] arrived while updating, adding it to batch of pending stops.", entityId)
+              entities.rememberingStop(entityId, Passivating)
+            case unexpected =>
+              log.warning("Got a terminated for [{}], entityId [{}] which is in state [{}]", ref, entityId, unexpected)
+          }
+        case OptionVal.None =>
+          // FIXME what would this Terminated be? Remember entities store stop?
+          stash()
+      }
+    case _: CoordinatorMessage         => stash()
+    case _: StartEntityInternal        => stash()
+    case _: RememberEntityCommand      => stash()
+    case l: LeaseLost                  => receiveLeaseLost(l)
+    case _: ShardRegion.StartEntityAck =>
+      // FIXME if we got an ack it was started elsewhere we should probably make sure we don't have it in current or pending starts?
+      stash()
+    case ShardRegion.Passivate(stopMessage) =>
+      if (VerboseDebug)
+        log.debug(
+          "Passivation of [{}] arrived while updating.",
+          entities.entityId(sender()).getOrElse(s"Unknown actor ${sender()}"))
+      passivate(sender(), stopMessage)
+    case _: ShardRegionCommand           => stash()
+    case msg: ShardQuery                 => receiveShardQuery(msg)
+    case PassivateIdleTick               => stash()
+    case msg: RememberEntityStoreCrashed => rememberEntityStoreCrashed(msg)
+    case msg if extractEntityId.isDefinedAt(msg) =>
+      deliverMessage(msg, sender())
+    case msg =>
+      // shouldn't be any other message types, but just in case
+      log.warning(
+        "Stashing unexpected message [{}] while waiting for remember entities update of starts [{}], stops [{}]",
+        msg.getClass,
+        update.started.mkString(", "),
+        update.stopped.mkString(", "))
+      stash()
+
+  }
+
+  def onUpdateDone(starts: Set[EntityId], stops: Set[EntityId]): Unit = {
+    // entities can contain both ids from start/stops and pending ones, so we need
+    // to mark the completed ones as complete to get the set of pending ones
+    starts.foreach { entityId =>
+      val stateBeforeStart = entities.entity(entityId).getOrElse(NoState)
+      // this will start the entity and transition the entity state in sessions to active
+      getOrCreateEntity(entityId)
+      sendMsgBuffer(entityId)
+      stateBeforeStart match {
+        case RememberingStart(Some(ackTo)) => ackTo ! ShardRegion.StartEntityAck(entityId, shardId)
+        case _                             =>
+      }
+      touchLastMessageTimestamp(entityId)
+    }
+    stops.foreach { entityId =>
+      // FIXME actually stop it by sendint stop message or what used to happen when stop write completed
```

Comment:
at the point haven't we already sent the stop as we are doing this after the termination has been received?

Reply:
I think I actually synced this up with the two end methods triggered by stop (`Passivation` and `StartEntityAck`), so I can remove the comment.